File: services/backend/src/schemas/cars.py
from typing import Optional, List, TypeVar, Generic

from pydantic import BaseModel
from pydantic.generics import GenericModel
from tortoise.contrib.pydantic import pydantic_model_creator

from src.database.models import Cars

T = TypeVar('T')

CarInSchema = pydantic_model_creator(Cars, name="CarIn", exclude = [
      "car_pictures", 
    ])
CarOutSchema = pydantic_model_creator(
    Cars, name="Car"
)

CarListSchema = pydantic_model_creator(Cars, name="CarList", include = [
    "id",
    "title",
    "ad_id",
    "url",
    "price",
    "make",
    "model",
    "color",
    "year",
    "city",
    "mileage",
    "hero_image",
    "seats",
    "doors",
    "color",
    "engine_size",
    "body_type",
    "fuel_type",
    "transmission"
])


class CarsFilter(GenericModel, Generic[T]):
    """Filtering"""
    page: int = 1
    limit: int = 10
    price: int
    year: int
    mileage: int
    gearbox: int
    engine_size: int
    colour: int
    fuel_type: int
    body_type: int
    doors: int
    seats: int
    make: str | None = None
    model: str | None = None


class PageResponse(GenericModel, Generic[T]):
    """ The response for a pagination query. """

    page_number: int
    page_size: int
    total_pages: int
    total_record: int
    content: List[CarListSchema]

File: services/frontend/src/store/modules/cars.js
import axios from 'axios';

const state = {
  cars: null,
  car: null
};

const getters = {
  stateCars: state => state.cars,
  stateCar: state => state.car,
};

const actions = {
  async getCars({commit}) {
    let {data} = await axios.get('cars');
    commit('setCars', data);
  },
  async filterCars({commit}) {
    let filterCarsPayload = {
      "page": 1,
      "limit": 200,
      "price": 0,
      "year": 0,
      "mileage": 0,
      "gearbox": 0,
      "engine_size": 0,
      "colour": 0,
      "fuel_type": 0,
      "body_type": 0,
      "doors": 0,
      "seats": 0,
      "make": "string",
      "model": "string"
    }
    let {data} = await axios.post('all', filterCarsPayload);
    commit('setCars', data.content);
  },
  async viewCar({commit}, id) {
    let {data} = await axios.get(`car/${id}`);
    commit('setCar', data);
  }
};

const mutations = {
  setCars(state, cars) {
    state.cars = cars;
  },
  setCar(state, car) {
    state.car = car;
  },
};

export default {
  state,
  getters,
  actions,
  mutations
};
